### ISSUE TITLE
`Duration.toString()` is wrong for many negative values

### ISSUE THREAD

**Issue by mernen** [dart-lang/sdk#48841]
The sign is completely dropped for durations `< Duration.zero && > Duration(hours: -1)`:

```dart
// Expected: -0:00:00.000123
// Actual:   0:00:00.000123
print(Duration(microseconds: -123));

// Expected: -0:30:00.000000
// Actual:   0:30:00.000000
print(Duration(minutes: -30));
```

Additionally, for certain values, Dart-JS will produce bogus sub-second representations. This is due to the lack of differentiation between `int` and `double` in dart2js and `num.remainder()` retaining the original receiver's sign, which ultimately culminates in `microseconds` having a value of `-0`:

```dart
// Expected: -1:00:00.000000
// Actual:   -1:00.00.00-0.0 (JS/DartPad only)
print(Duration(hours: -1));
```

[Demo DartPad](https://dartpad.dev/?id=e792348559e89e20fdf338aa98690cef)

Tested on:

* `Dart SDK version: 2.16.2 (stable) (Tue Mar 22 13:15:13 2022 +0100) on "macos_x64"`
* dartpad.dev "Based on Flutter 2.10.5 Dart SDK 2.16.2", on Chrome 100

My proposal is to change `Duration.toString()` by moving the negative `microseconds` test up, and having an explicit sign rather than relying on `hours` becoming negative: (apologies for not preparing a proper patch)

```diff
diff --git i/sdk/lib/core/duration.dart w/sdk/lib/core/duration.dart
index 11fc5c83db6..d3b827e47bb 100644
--- i/sdk/lib/core/duration.dart
+++ w/sdk/lib/core/duration.dart
@@ -330,13 +330,16 @@ class Duration implements Comparable<Duration> {
   /// print(d.toString()); // 1:10:00.000500
   /// ```
   String toString() {
+    var sign = "";
     var microseconds = inMicroseconds;
+    if (microseconds < 0) {
+      sign = "-";
+      microseconds = -microseconds;
+    }
 
     var hours = microseconds ~/ microsecondsPerHour;
     microseconds = microseconds.remainder(microsecondsPerHour);
 
-    if (microseconds < 0) microseconds = -microseconds;
-
     var minutes = microseconds ~/ microsecondsPerMinute;
     microseconds = microseconds.remainder(microsecondsPerMinute);
 
@@ -348,7 +351,7 @@ class Duration implements Comparable<Duration> {
     var secondsPadding = seconds < 10 ? "0" : "";
 
     var paddedMicroseconds = microseconds.toString().padLeft(6, "0");
-    return "$hours:"
+    return "$sign$hours:"
         "$minutesPadding$minutes:"
         "$secondsPadding$seconds.$paddedMicroseconds";
   }
```

**Comment by lrhn** [comment]
Pretty good solution, only fails for the minimal integer (and only on native, not web).

**Comment by mernen** [comment]
You're right, that darned minimal is a pain that I hadn't considered...

**Comment by rakudrama** [comment]
On the web platforms you can avoid `-0.0` by using subtraction from zero rather than negation.
```
microseconds = 0 - microseconds;
```
Be sure to add a test for this case, as it would be tempting to remove the `0 `.

**Comment by lrhn** [comment]
The fix ensures that `Duration(-0.0)` is treated as `Duration(0)`, which is consistent with the internal state being an `int`, and `-0.0` not being an `int` (but if it is an `int`, it's `0`).

You cannot have a negative zero-length duration. (Well, I guess you can, but you can't easily distinguish it from a non-negative zero-length duration, which is the intended behavior).

**Comment by mernen** [comment]
My point is that this test the way it's written doesn't trigger the wrong behavior; a more appropriate test would be for `Duration(hours: -1)`, which is still broken.

That being said, I believe the better approach would be to take the -0 test out, fix this particular issue at the root in dart2js (#48884), and then consider this bug completely fixed.
